### PR TITLE
Add script termination option to 'Lua Scripts' floater

### PR DESCRIPTION
**Pull request by maxim-productengine**
Add the way to terminate selected script in 'Lua Scripts' floater: call `lluau::error` when it's needed (but if the script is blocked in `getNext()`, close the queue first).

### CODE REVIEW

**Comment by nat-goodspeed** [comment]
Maxim, thanks, I'm looking, but it will probably be tomorrow before I can respond.

File: indra/newview/llluamanager.cpp [hunk 1]
```diff
@@ -188,6 +189,20 @@ void LLLUAmanager::runScriptFile(const std::string &filename, script_result_fn r
             // A script_finished_fn is used to initialize the LuaState.
             // It will be called when the LuaState is destroyed.
             LuaState L(finished_cb);
+
+            lua_callbacks(L)->interrupt = [](lua_State *L, int gc)
+            {
+                if (gc >= 0)
+                    return;
```
Comment (nat-goodspeed) [inline indra/newview/llluamanager.cpp:196]: A comment explaining the meaning of the `gc` parameter would be helpful here.

File: indra/newview/llluamanager.cpp [hunk 2]
```diff
@@ -188,6 +189,20 @@ void LLLUAmanager::runScriptFile(const std::string &filename, script_result_fn r
...
+               
+                std::set<std::string> scripts = LLLUAmanager::getTerminationList();
+                std::string coro = LLCoros::getName();
+                if (scripts.find(coro) != scripts.end()) 
+                {
+                    sTerminationList.erase(coro);
```
Comment (nat-goodspeed) [inline indra/newview/llluamanager.cpp:202]: Maybe capture the iterator returned by `find()` and `erase()` that?

Oh! I don't think this `erase()` will actually affect `sTerminationList`! `getTerminationList()` returns `std::set<std::string>` by value, so in this function `scripts` is a _copy_ of `sTerminationList`!

Maybe `runScriptFile()` could directly reference `sTerminationList`?

**Review by nat-goodspeed — CHANGES_REQUESTED** [review]
I know you're still working on this, so maybe my comments have become moot. But there are a couple things in this PR as it now stands that I'd like to see changed.

File: indra/newview/llfloaterluascripts.cpp [hunk 3]
```diff
@@ -113,6 +118,7 @@ void LLFloaterLUAScripts::onScrollListRightClicked(LLUICtrl *ctrl, S32 x, S32 y)
         if (menu)
         {
             mTargetFolderPath = std::filesystem::path((item->getColumn(1)->getValue().asString())).parent_path().string();
+            mCoroName = item->getValue().asString();
```
[the remark this answers is not in the document]
Comment (maxim-productengine) [reply indra/newview/llfloaterluascripts.cpp]: Yes, that what we usually use for the scroll list menu.
I got rid of mTargetFolderPath and mCoroName, and moved appropriate code to the callback part.

File: indra/llcommon/llapp.cpp [hunk 4]
```diff
@@ -429,6 +429,7 @@ void LLApp::setStatus(EAppStatus status)
             statsd = LLSD::Integer(status);
         }
         LLEventPumps::instance().obtain("LLApp").post(llsd::map("status", statsd));
+        LLEventPumps::instance().obtain("LLLua").post(llsd::map("status", "close_all"));
```
[the remark this answers is not in the document]
Comment (maxim-productengine) [reply indra/llcommon/llapp.cpp:432]: To be honest I missed the ability to terminate coroutine once or twice in the past when working on maint bugs. So, I'm for it.

I made changes according to your comments. And I'm not planning any more changes for now for this PR. (there will be some changes related to script termination as you know, but I'll create separate pr)

**Review by nat-goodspeed — APPROVED** [review]
Thanks, Maxim!

File: indra/llcommon/llapp.cpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (nat-goodspeed) [reply indra/llcommon/llapp.cpp:432]: I'm sorry to say I overlooked an important point. I can indeed introduce a flag for the next time a given coroutine wakes up and calls `LLCoros::checkStop()`, but I can't cause that coroutine to wake up and check it. An `LLEventStream` listener runs on the coroutine calling `post()`. An exception thrown by such a listener will propagate through the `post()` call.

The Boost.Fiber library we use does not provide a way to force an exception onto a suspended fiber -- although tantalizingly, the Boost.Context library on which it's based provides a mechanism that could support that. At present, the best we can do is to wait for the subject coroutine to call `checkStop()`.

**Pull request merged**